server/events.go

```diff
@@ -0,0 +1,271 @@
+
+	s.stampServerInfo(&m.Server)
+	msg, _ := json.MarshalIndent(m, "", "  ")
+	s.sendInternalMsg(r, subj, msg)
+}
+
+// accountDisconnectEvent will send an account client disconnect event if there is interest.
+func (s *Server) accountDisconnectEvent(c *client, now time.Time, reason string) {
+	if s.sys == nil || s.sys.client == nil || s.sys.account == nil {
+		return
+	}
+	acc := s.sys.account
+
+	subj := fmt.Sprintf(disconnectEventSubj, c.acc.Name)
+	r := acc.sl.Match(subj)
+	if s.noOutSideInterest(r) {
+		return
+	}
+
+	c.mu.Lock()
+	m := DisconnectEventMsg{
+		Client: ClientInfo{
+			Start:   c.start,
+			Stop:    &now,
+			Host:    c.host,
+			ID:      c.cid,
+			Account: c.acc.Name,
+			User:    nameForClient(c),
+			Name:    c.opts.Name,
+			Lang:    c.opts.Lang,
+			Version: c.opts.Version,
+		},
+		Sent: DataStats{
+			Msgs:  c.inMsgs,
+			Bytes: c.inBytes,
+		},
+		Received: DataStats{
+			Msgs:  c.outMsgs,
+			Bytes: c.outBytes,
+		},
+		Reason: reason,
+	}
+	c.mu.Unlock()
+
+	s.stampServerInfo(&m.Server)
+	msg, _ := json.MarshalIndent(m, "", "  ")
+	s.sendInternalMsg(r, subj, msg)
+}
+
+// Internal message callback. If the msg is needed past the callback it is
+// required to be copied.
+type msgHandler func(sub *subscription, subject, reply string, msg []byte)
+
+func (s *Server) deliverInternalMsg(sub *subscription, subject, reply, msg []byte) {
+	if s.sys == nil {
+		return
+	}
+	s.mu.Lock()
+	cb := s.sys.subs[string(sub.sid)]
+	s.mu.Unlock()
+	if cb != nil {
+		cb(sub, string(subject), string(reply), msg)
+	}
+}
+
+// Create an internal subscription. No support for queue groups atm.
+func (s *Server) sysSubscribe(subject string, cb msgHandler) (*subscription, error) {
+	if s.sys == nil {
+		return nil, ErrNoSysAccount
+	}
+	if cb == nil {
+		return nil, fmt.Errorf("Undefined message handler")
+	}
+	s.mu.Lock()
+	sid := strconv.FormatInt(int64(s.sys.sid), 10)
+	s.sys.subs[sid] = cb
+	s.sys.sid++
+	c := s.sys.client
+	s.mu.Unlock()
+
+	// Now create the subscription
+	if err := c.processSub([]byte(subject + " " + sid)); err != nil {
+		return nil, err
+	}
+	c.mu.Lock()
+	sub := c.subs[sid]
+	c.mu.Unlock()
+	return sub, nil
+}
+
+func (s *Server) sysUnsubscribe(sub *subscription) {
+	if sub == nil || s.sys == nil {
+		return
+	}
+	s.mu.Lock()
+	acc := s.sys.account
+	c := s.sys.client
+	delete(s.sys.subs, string(sub.sid))
+	s.mu.Unlock()
+	c.unsubscribe(acc, sub, true)
+}
+
+func (s *Server) noOutSideInterest(r *SublistResult) bool {
+	sc := s.sys.client
+	if sc == nil || r == nil {
+		return true
+	}
+	nsubs := len(r.psubs) + len(r.qsubs)
+	if nsubs == 0 {
+		return true
+	}
+	// We will always be no-echo but will determine that on delivery.
+	// Here we try to avoid generating the payload if there is only us.
+	// We only check normal subs. If we introduce queue subs into the
+	// internal subscribers we should add in the check.
+	for _, sub := range r.psubs {
+		if sub.client != sc {
+			return false
+		}
+	}
+	return false
```

Comment:
Should it be `return true`? to indicate no outside interest? (since the only sub we found was for the internal client)

Reply:
Correct.. Double negatives.. ;)